modules/nodesale/processor.go

```diff
@@ -0,0 +1,273 @@
+	}, nil
+}
+
+// Name implements indexer.Processor.
+func (p *Processor) Name() string {
+	return "nodesale"
+}
+
+func extractNodesaleData(witness [][]byte) (data []byte, internalPubkey *btcec.PublicKey, isNodesale bool) {
+	tokenizer, controlBlock, isTapScript := extractTapScript(witness)
+	if !isTapScript {
+		return []byte{}, nil, false
+	}
+	state := 0
+	for tokenizer.Next() {
+		switch state {
+		case 0:
+			if tokenizer.Opcode() == txscript.OP_0 {
+				state++
+			} else {
+				state = 0
+			}
+		case 1:
+			if tokenizer.Opcode() == txscript.OP_IF {
+				state++
+			} else {
+				state = 0
+			}
+		case 2:
+			if tokenizer.Opcode() == txscript.OP_DATA_4 &&
+				bytes.Equal(tokenizer.Data(), NODESALE_MAGIC) {
+				state++
+			} else {
+				state = 0
+			}
+		case 3:
+			if tokenizer.Opcode() == txscript.OP_PUSHDATA1 ||
+				tokenizer.Opcode() == txscript.OP_PUSHDATA2 ||
+				tokenizer.Opcode() == txscript.OP_PUSHDATA4 {
+				data := tokenizer.Data()
+				return data, controlBlock.InternalKey, true
+			}
+			state = 0
+		}
+	}
+	return []byte{}, nil, false
+}
+
+type nodesaleEvent struct {
+	transaction  *types.Transaction
+	eventMessage *protobuf.NodeSaleEvent
+	eventJson    []byte
+	txPubkey     *btcec.PublicKey
+	rawData      []byte
+}
+
+func (p *Processor) parseTransactions(ctx context.Context, transactions []*types.Transaction) ([]nodesaleEvent, error) {
+	var events []nodesaleEvent
+	for _, t := range transactions {
+		for _, txIn := range t.TxIn {
+			data, txPubkey, isNodesale := extractNodesaleData(txIn.Witness)
+			if !isNodesale {
+				continue
+			}
+
+			event := &protobuf.NodeSaleEvent{}
+			err := proto.Unmarshal(data, event)
+			if err != nil {
+				logger.WarnContext(ctx, "Invalid Protobuf",
+					slogx.String("block_hash", t.BlockHash.String()),
+					slogx.Int("txIndex", int(t.Index)))
+				continue
+			}
+			eventJson, err := protojson.Marshal(event)
+			if err != nil {
+				return []nodesaleEvent{}, errors.Wrap(err, "Failed to parse protobuf to json")
+			}
+
+			events = append(events, nodesaleEvent{
+				transaction:  t,
+				eventMessage: event,
+				eventJson:    eventJson,
+				rawData:      data,
+				txPubkey:     txPubkey,
+			})
+		}
+	}
+	return events, nil
+}
+
+// Process implements indexer.Processor.
+func (p *Processor) Process(ctx context.Context, inputs []*types.Block) error {
+	for _, block := range inputs {
+		logger.InfoContext(ctx, "Nodesale processing a block",
+			slogx.Int64("block", block.Header.Height),
+			slogx.Stringer("hash", block.Header.Hash))
+		// parse all event from each transaction including reading tx wallet
+		events, err := p.parseTransactions(ctx, block.Transactions)
+		if err != nil {
+			return errors.Wrap(err, "Invalid data from bitcoin client")
+		}
+		// open transaction
+		qtx, err := p.datagateway.BeginNodesaleTx(ctx)
+		if err != nil {
+			return errors.Wrap(err, "Failed to create transaction")
+		}
+		defer func() {
+			err = qtx.Rollback(ctx)
```

Comment:
`err = qtx.Rollback(ctx)` does nothing as it is not the return value.

Either set the `err` as return variable in the function return value definition OR handle the error right here if it is not nil.